### PR TITLE
fix: add graphql query to display slugs on the completed challenges in the profile

### LINKED ISSUE #36085: Broken links on profile page
**Describe your problem and how to reproduce it:**
the challenge links in the profile timeline are broken.

image

### PR DESCRIPTION
**Pull request by ahmaxed**
- [x] I have read [freeCodeCamp's contribution guidelines](https://github.com/freeCodeCamp/freeCodeCamp/blob/master/CONTRIBUTING.md).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `master` branch of freeCodeCamp.
- [x] None of my changes are plagiarized from another source without proper attribution.
- [x] All the files I changed are in the same world language (for example: only English changes, or only Chinese changes, etc.)
- [x] My changes do not use shortened URLs or affiliate links.

closes #36085

### CODE REVIEW

File: client/src/components/profile/Profile.js [hunk 1]
```diff
@@ -128,43 +128,73 @@ function Profile({ user, isSessionUser }) {
...
+              }
+            }
+          }
+        }
+      `}
+      render={data => {
+        const idToSlugMap = data.allChallengeNode.edges.reduce(
+          (a, b) => Object.assign(a, { [b.node.id]: b.node.fields.slug }),
+          {}
+        );
```
Comment (ahmadabdolsaheb) [inline client/src/components/profile/Profile.js]: I do not love how idToSlugMap is made. there must be a better way using graphql. please let me know if anything comes to mind.

File: client/src/components/profile/components/TimeLine.js [hunk 2]
```diff
@@ -68,15 +69,19 @@ class Timeline extends Component {
...
-    const { idToNameMap } = this.props;
+    const { idToNameMap, idToSlugMap } = this.props;
     const { id, completedDate } = completed;
     const challengeDashedName = idToNameMap[id];
+    let slug = `/challenges/${challengeDashedName}`;
+
+    if (idToSlugMap[id]) {
+      slug = idToSlugMap[id];
+    }
+
```
Comment (ahmadabdolsaheb) [inline client/src/components/profile/components/TimeLine.js]: Also, if the challenge id is not found in idToSlugMap, I set the slug to be what it was before so debugging would be easier by seeing the name of the challenge in the url for making and resolving issues.

Let me know if something better comes to mind. 
a flash message probably?

File: client/src/components/profile/components/TimeLine.js [hunk 3]
```diff
@@ -68,15 +69,19 @@ class Timeline extends Component {
...
   renderCompletion(completed) {
-    const { idToNameMap } = this.props;
+    const { idToNameMap, idToSlugMap } = this.props;
     const { id, completedDate } = completed;
     const challengeDashedName = idToNameMap[id];
+    let slug = `/challenges/${challengeDashedName}`;
+
+    if (idToSlugMap[id]) {
+      slug = idToSlugMap[id];
+    }
```
Comment (ahmadabdolsaheb) [inline client/src/components/profile/components/TimeLine.js]: Also, if the challenge id is not found in idToSlugMap, I set the slug to be what it was before so debugging would be easier by seeing the name of the challenge in the url for making and resolving issues.

Let me know if something better comes to mind.
a flash message probably?

File: client/src/components/profile/components/TimeLine.js [hunk 4]
```diff
@@ -68,15 +69,19 @@ class Timeline extends Component {
   }
 
   renderCompletion(completed) {
-    const { idToNameMap } = this.props;
+    const { idToNameMap, idToSlugMap } = this.props;
     const { id, completedDate } = completed;
     const challengeDashedName = idToNameMap[id];
+    let slug = `/challenges/${challengeDashedName}`;
+
+    if (idToSlugMap[id]) {
```
Comment (ahmadabdolsaheb) [inline client/src/components/profile/components/TimeLine.js]: Also, if the challenge id is not found in idToSlugMap, I set the slug to be what it was before so debugging would be easier by seeing the name of the challenge in the url for making and resolving issues.

Let me know if something better comes to mind.
a flash message probably?

File: client/src/components/profile/components/TimeLine.js [hunk 5]
```diff
@@ -68,15 +69,19 @@ class Timeline extends Component {
...
+      slug = idToSlugMap[id];
```
Comment (ahmadabdolsaheb) [inline client/src/components/profile/components/TimeLine.js]: Also, if the challenge id is not found in idToSlugMap, I set the slug to be what it was before so debugging would be easier by seeing the name of the challenge in the url for making and resolving issues.

Let me know if something better comes to mind.
a flash message probably?

**Comment by ojeytonwilliams** [comment]
@ahmadabdolsaheb I may be missing something here, but isn't the root problem that the variable called `challengeDashedName` is not actually the dashed name?  If you `dasherize` that, the resulting links should work.

It does mean the api server gets pinged.  Was that something you were trying to avoid?

**Comment by raisedadead** [comment]
Bump.

**Comment by raisedadead** [comment]
Tagged the original issue as must fix in the project board instead of the PR. Apologies about the confusion.

**Comment by ahmadabdolsaheb** [comment]
@ojeytonwilliams thank you for the review and apologies for the long delay.

Currently  /challenges in the url resolves to 404 locally and .dev but not on production.
For example:

http://localhost:8000/challenges/headline-with-the-h2-element
resolves to 404

While
https://freecodecamp.org/challenges/headline-with-the-h2-element
resolves to
https://learn.freecodecamp.org/responsive-web-design/basic-html-and-html5/headline-with-the-h2-element/

Let me know if anything comes to your mind. Otherwise I will need to investigate it a little bit further.

**Comment by ojeytonwilliams** [comment]
No worries about the delay.

You're quite right, it does result in a 404, but I don't *think* it did at the time I reviewed this.  Regardless, what I've since found is that http://localhost:3000/challenges/headline-with-the-h2-element does work locally. So, it should work if you use `apiLocation` from `env.json`.  Something like:

```js
`${apiLocation}/challenges/${dasherize(challengeDashedName)}`
```
works on my machine.

This does, obviously, mean we'd be pinging the server a bunch of times, so maybe this isn't so smart!

**Comment by ahmadabdolsaheb** [comment]
@ojeytonwilliams thanks for confirming that.
That does seem like a simple fix. However, let me know if anything comes to mind before finalizing this solution. thank you.

**Comment by ojeytonwilliams** [comment]
Maybe we're going about this the wrong way.  Wouldn't it be better to have redirects from all the `/challenges/<id or short name>/` urls to the curriculum?  If we create those as we create the curriculum we can avoid having to search the ids and also avoid pinging the server each time.

I have in mind a modification to `gatsby-node` using something like
```js
result.data.allChallengeNode.edges.forEach(challenge =>
  createChallengeRedirect(challenge);
  createChallengePages(createPage)(challenge);
);
```

Then the profile would just use those redirects.  What do you think, @ahmadabdolsaheb?

**Comment by raisedadead** [comment]
Hey @moT01 and @ojeytonwilliams I am going to close this in favor of Tom's PR.

**Pull request closed without merging**